### PR TITLE
Add GPT2 example for Sagamaker distributed modelparallel

### PR DESCRIPTION
**Pull request by YangFei1990**
*Issue #, if available:*

*Description of changes:*
- Adding a GPT2 example
- Removing the old mnist example as the model itself is too small and won't be useful for customers.
- Clear a minor bug for the Bert example

Dataset used:
- Wiki-text
- Openwebtext
Both datasets have been legal reviewed.

*Testing done:*
End-to-end test of the bert and GPT2 notebook
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `tox -e black-format,black-nb-format`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/distributed_training/pytorch/model_parallel/gpt2/data_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            data = OpenwebtextPretrainingDataset(
+                input_paths=input_paths, max_sequence_length=max_sequence_length, zipped=zipped, use_last_file_only=use_last_file_only
+            )
+        elif data_type == "wiki":
+            if len(input_paths) > 1:
+                print(f"Wiki data only support single file when calling create_pretraining_dataloader, reading the first file instead..")
+            data = WikiPretrainingDataset(input_file=input_paths[0], max_pred_length=max_sequence_length)
+        else:
+            raise ValueError(f"Unsupported data type {data_type}")
+        # TODO: set sampler.epoch to correctly shuffle across epochs, else same order will be used for all epochs
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/data_pipeline.py:146]: If this TODO item is not relevant, can it be removed?

File: training/distributed_training/pytorch/model_parallel/gpt2/data_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        padded_mask_indices = (masked_lm_positions == 0).nonzero(as_tuple=False)
+        if len(padded_mask_indices) != 0:
+            index = padded_mask_indices[0].item()
+        masked_lm_labels[masked_lm_positions[:index]] = masked_lm_ids[:index]
+
+        return [input_ids, segment_ids, input_mask, masked_lm_labels, next_sentence_labels]
+
+
+###### Load Openwebtext pretraining data ######
+class OpenwebtextPretrainingDataset(torch.utils.data.Dataset):
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/data_pipeline.py:60]: Please add a list in the PR description of all datasets being used in this example. Each one must be reviewed and approved.

File: training/distributed_training/pytorch/model_parallel/gpt2/data_prep_512.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import torch
+from datasets import load_dataset
+from transformers import GPT2TokenizerFast
+# download the unprocessed dataset
+dataset = load_dataset('openwebtext', split='train')
+tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")
+
+# Process the dataset and split it into train and test subsets
+dataset = dataset.map(lambda e: tokenizer(e['text'], max_length=512, truncation=True), num_proc=96)
+print(dataset)
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/data_prep_512.py:13]: There are tons of print statements - can most/all be removed?

File: training/distributed_training/pytorch/model_parallel/gpt2/fp16/__init__.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# coding=utf-8
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/fp16/__init__.py:2]: This file is copyright NVIDA - What is the original source of this file and how has it been modified? (Applies to at least 5 files in this PR)

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 5]
```diff
@@ -0,0 +1,1166 @@
...
+            break
+
+        del train_dataloader
+
+        if args.parallel_proc_data_processing:
+            s = time.time()
+            train_dataloader = dataset_future.result(timeout=None)
+            wait_time = time.time() - s
+            if wait_time > 1:
+                # TODO if this happens, we should try num_workers>1 in dataloader
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:675]: Address and remove all TODO items.

**Comment by jkroll-aws** [comment]
What is the reason for removing the MNIST example?

File: training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb [hunk 6]
```diff
@@ -0,0 +1,683 @@
+{
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Grammar - pluralize - <code>the following files/folders</code></p><p> </p><p>Be consistent with use of period or no period at end of list lines.</p><p> </p><p>Prerequisites - use the default SageMaker S3 bucket. No need to create a new one. This is what your code is already doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #12.    print(f'SageMaker Execution Role:{role}')
</pre><p>Nit - add spaces before all formatted strings for easier readability, ie <code>Role: {role}</code>. 3 occurrences.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Clarify what CI means/stands for, for readers who are unfamiliar.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Use an approved example bucket name, <code>DOC-EXAMPLE-BUCKET</code>	, rather than "your-bucket".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Grammar/branding - change <code>fsx</code> and <code>FSX</code> to standard spelling <code>FSx</code> in all text/markdown and section titles.</p><p> </p><p>Use page title rather than "here" as link text - <code>Distributed Training of Mask-RCNN in Amazon SageMaker using FSx</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        file_system_id = "fs-01bdca7270e7a4c38"
</pre><p>Remove hard-coded file system ID, security group ID, subnet. Replace with something like <code>"&lt;your-file-system-id&gt;"</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Grammar - pluralize "instances" in equation - <code>the number of instances</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    base_job_name = f'smp-{model_config}-{machine_str}-tp{tp_degree}-pp{pp_degree}-bs{hyperparameters["train_batch_size"]}'
</pre><p>Add a print statement so it's clear what the base job name is from concatenating all these different values?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #43.            image_uri=f"855988369404.dkr.ecr.{region}.amazonaws.com/fewu-test:pytorch_1.7_tp_efa",
</pre><p>Update the image URI when it is available in a different location.</p><p> </p><p>Note that this will not work in China regions, which use <code>amazonaws.com.cn</code>  .</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>Branding - capitalize <code>SageMaker Batch Transform</code>.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

File: training/distributed_training/pytorch/model_parallel/gpt2/data_prep_512.py (hunk 3 above)
Comment (YangFei1990) [reply training/distributed_training/pytorch/model_parallel/gpt2/data_prep_512.py]: Those prints are helpful for user to understand the current progress, I would recommend to keep them.

File: training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb (hunk 6 above)
Comment (YangFei1990) [reply training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb]: User needs to create a S3 bucket for their own data. The notebook is using the default SageMaker S3 bucket with CI data only.

**Comment by YangFei1990** [comment]
> What is the reason for removing the MNIST example?

Updated reason in the description.

File: training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb (hunk 6 above)
Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>fixed</p>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/submit-train-gpt-simple.ipynb:1]: <p>It seems the description is not perfectly matching with the following code. Updating the paragraphs.</p>

**Review by mchoi8739 — DISMISSED** [review]

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 7]
```diff
@@ -0,0 +1,716 @@
+{
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p><code>separate EC2 instance/notebook</code>  -&gt; Clarify <code>separate EC2 instance or SageMaker notebook</code>  ?</p><p>Typo - <code>amazon-sagemaker-examples</code>  repository (missing plural "s" on examples)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>Nit - two mentions of "for CI" - an external user will not know what that means. Elaborate a bit to something like "for continuous integration testing"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.        # Donwload some mock data from a public bucket in us-east-1
</pre><p>Typo <code>Donwload</code> -&gt; <code>Download</code> ﻿ </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>Branding nit - capitalize <code>Lustre</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>Nit - "requires a small memory" sounds odd to me - maybe "requires a small amount of memory", or be more specific, like "typically requires no more than a few MB of memory"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3034/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: Choosing "a SageMaker notebook instance"

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>Also I mentioned "continuous integration (CI) test" in this markdown cell.</p>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>See my comment in the previous section.</p>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>fixed</p>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>fixed all 3 instances</p>

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:1]: <p>Thanks for suggestions. Using "requires a small amount of memory".</p><p> </p>

**Review by mchoi8739 — APPROVED** [review]

**Review by jkroll-aws — APPROVED** [review]
